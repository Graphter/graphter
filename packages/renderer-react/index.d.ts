import ListRenderer from './list-renderer';
export { ListRenderer };
