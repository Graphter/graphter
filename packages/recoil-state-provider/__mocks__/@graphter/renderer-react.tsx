import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)