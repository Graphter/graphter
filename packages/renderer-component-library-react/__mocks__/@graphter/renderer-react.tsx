import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()