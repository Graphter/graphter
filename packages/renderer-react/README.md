# Graphter React renderer

A React framework that, when combined with node renderers such as those provided in the [Component Lib](../renderer-component-library-react/README.md) package, allows humans to manage arbitrary data structures.