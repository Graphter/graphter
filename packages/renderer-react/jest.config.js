module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  moduleNameMapper: {
    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.ts",
    "\\.(pcss|css|less)$": "identity-obj-proxy"
  },
  setupFilesAfterEnv: [ './setupTests.ts' ],
  automock: true,
  unmockedModulePathPatterns: [],
  snapshotSerializers: [
    'enzyme-to-json'
  ],
  testMatch: ['<rootDir>/**/*.test.ts*'],
  testPathIgnorePatterns: ['node_modules'],
};