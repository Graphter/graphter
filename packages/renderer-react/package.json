{
  "name": "malle-renderer-react",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "cross-env CI=1 react-scripts test --env=jsdom",
    "test:watch": "react-scripts test --env=jsdom",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "dependencies": {},
  "devDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-react": "^7.9.1",
    "@babel/preset-stage-0": "^7.8.3",
    "@svgr/rollup": "^5.3.0",
    "@types/react": "^16.9.25",
    "eslint": "^6.8.0",
    "react-scripts": "^3.4.1",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-postcss": "^2.5.0",
    "rollup-plugin-typescript2": "^0.26.0",
    "rollup-plugin-url": "^3.0.1",
    "typescript": "^3.8.3"
  }
}
