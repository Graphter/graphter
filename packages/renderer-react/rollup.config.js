import babel from 'rollup-plugin-babel'
import commonjs from 'rollup-plugin-commonjs'
import external from 'rollup-plugin-peer-deps-external'
import postcss from 'rollup-plugin-postcss'
import resolve from 'rollup-plugin-node-resolve'
import url from 'rollup-plugin-url'
import typescriptPlugin from 'rollup-plugin-typescript2'
import typescript from 'typescript';
import svgr from '@svgr/rollup'

import pkg from './package.json'

const externals = [
  ...Object.keys(pkg.peerDependencies),
  ...Object.keys(pkg.dependencies)
];
console.log('externals', externals);

export default {
  input: 'src/index.ts',
  output: [
    {
      file: pkg.main,
      format: 'cjs',
      sourcemap: true
    },
    {
      file: pkg.module,
      format: 'es',
      sourcemap: true
    }
  ],
  plugins: [
    // For debugging: { load(id) {console.log(id)} },
    external(externals),
    typescriptPlugin({
      typescript
    }),
    postcss({
      modules: true
    }),
    url(),
    svgr(),
    babel({
      exclude: 'node_modules/**',
      plugins: [ '@babel/plugin-proposal-export-default-from' ]
    }),
    resolve(),
    commonjs({
      namedExports: {
        'recoil': [ 'RecoilRoot', 'useRecoilState', 'useRecoilValue', 'useRecoilValueLoadable', 'useRecoilCallback', 'atom', 'selector' ]
      }
    })
  ]
}
