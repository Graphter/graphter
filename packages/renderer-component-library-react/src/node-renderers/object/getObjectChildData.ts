import { GetChildDataFn } from "@graphter/core";
import { nodeRendererStore } from "@graphter/renderer-react";

export const getObjectChildData: GetChildDataFn = (config, path, getNodeValue) => {
  if (!config.children?.length) throw new Error(`${config.type} type '${config.id}' has no children configured. At least one is required.`)

  const childValues = config.children.map(childConfig => {
    const childPath = [ ...path, childConfig.id ]
    const childRenderer = nodeRendererStore.get(childConfig.type)
    const nodeData = childRenderer.getChildData ?
      childRenderer.getChildData(childConfig, childPath, getNodeValue) :
      getNodeValue(childPath)
    return { key: childConfig.id, value: nodeData }
  })
  const reFormedObject = childValues.reduce<{ [key: string]: any }>((a, c) => {
    a[c.key] = c.value
    return a
  }, {})
  console.info(`Reformed ${childValues.length} child values for node '${config.id}'`, reFormedObject)
  return reFormedObject
}