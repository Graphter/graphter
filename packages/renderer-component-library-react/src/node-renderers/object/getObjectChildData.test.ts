import { pathConfigStore, nodeRendererStore } from "@graphter/renderer-react";
import { when } from "jest-when";
import { getObjectChildData } from "./getObjectChildData";

const nodeRendererStoreMock  = nodeRendererStore as jest.Mocked<any>
const pathConfigStoreMock = pathConfigStore as jest.Mocked<any>

describe('getChildData()', () => {
  it('should return descendent data correctly formed', () => {
    pathConfigStoreMock.get.mockReturnValueOnce({
      id: 'page',
      type: 'object',
      children: [
        {
          id: 'author',
          type: 'object',
          children: [ // Just here for illustration; don't actually do anything in the test
            { id: 'name', type: 'string' },
            { id: 'location', type: 'string' }
          ]
        }
      ]
    })
    when(nodeRendererStoreMock.get)
      .calledWith('object')
      .mockReturnValueOnce({
        getChildData: () => ({ name: 'Bob', location: 'London' })
      })
    const result = getObjectChildData(['page'], jest.fn())
    expect(result).toEqual({
      author: { name: 'Bob', location: 'London' }
    })
  })
  it('should error if parent config is not found', () => {
    pathConfigStoreMock.get.mockReturnValue(null)
    expect(() => getObjectChildData(['page'], jest.fn()))
      .toThrowErrorMatchingSnapshot()
  })
  it.each([[], null])
    ('should error if the parent has no children (i.e. %s) configured', (noChildren) => {
    pathConfigStoreMock.get.mockReturnValue({
      id: 'page',
      type: 'object',
      children: noChildren
    })
    when(nodeRendererStoreMock.get)
      .calledWith('object')
      .mockReturnValueOnce({
        getChildData: () => ({ name: 'Bob', location: 'London' })
      })
    expect(() => getObjectChildData(['page'], jest.fn()))
      .toThrowErrorMatchingSnapshot()
  })
  it('should skip descendent data resolution when child renderer does not implement a getChild() function', () => {
    const getNodeValueMock = jest.fn()
    pathConfigStoreMock.get.mockReturnValueOnce({
      id: 'page',
      type: 'object',
      children: [
        { id: 'title', type: 'string' },
        { id: 'author', type: 'string' }
      ]
    })
    when(nodeRendererStoreMock.get)
      .calledWith('string')
      .mockReturnValue({ })
    when(getNodeValueMock)
      .calledWith(['page', 'title'])
      .mockReturnValueOnce('The Page Title')
      .calledWith(['page', 'author'])
      .mockReturnValueOnce('Joe Bloggs')
    const result = getObjectChildData(['page'], getNodeValueMock)
    expect(result).toEqual({
      title: 'The Page Title',
      author: 'Joe Bloggs'
    })
  })
})