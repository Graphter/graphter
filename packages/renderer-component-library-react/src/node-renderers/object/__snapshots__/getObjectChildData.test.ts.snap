// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChildData() should error if null parent config is not found 1`] = `"Cannot read property 'children' of null"`;

exports[`getChildData() should error if parent config is not found 1`] = `"Couldn't find config for node at path 'page'"`;

exports[`getChildData() should error if the parent has no children (i.e. []) configured 1`] = `"object type 'page' has no children configured. At least one is required."`;

exports[`getChildData() should error if the parent has no children (i.e. null) configured 1`] = `"object type 'page' has no children configured. At least one is required."`;

exports[`getChildData() should error if undefined parent config is not found 1`] = `"Cannot read property 'children' of undefined"`;
